File: de.unistuttgart.iste.sqa.mpw.karasimulator.java/core/src/main/java/de/unistuttgart/kara/commands/impl/CompositeCommandBase.java
package de.unistuttgart.kara.commands.impl;

import de.unistuttgart.kara.basetypes.Entity;
import de.unistuttgart.kara.commands.Command;
import de.unistuttgart.kara.commands.CompositeCommand;

public abstract class CompositeCommandBase extends CompositeCommand {

    @Override
    public void undo() {
        var subCommands = getSubCommands();
        for (int i = subCommands.size() - 1; i >= 0; i--) {
            Command subCommand = subCommands.get(i);
            subCommand.undo();
        }
    }

    @Override
    public void redo() {
        var subCommands = getSubCommands();
        for (Command subCommand : subCommands) {
            subCommand.redo();
        }
    }

    protected void executeSetProperty(Entity entity, String propertyName, Object oldValue, Object newValue) {
        var command = new de.unistuttgart.kara.commands.impl.SetPropertyCommandImpl();
        command.setEntity(entity);
        command.setOldValue(oldValue);
        command.setNewValue(newValue);
        command.setPropertyName(propertyName);

        addToSubCommands(command);
        command.execute();
    }

    protected void executeAddReference(Entity entity, String propertyName, Entity entityToAdd) {
        var command = new de.unistuttgart.kara.commands.impl.AddEntityCommandImpl();
        command.setEntity(entity);
        command.setEntityToAdd(entityToAdd);
        command.setPropertyName(propertyName);

        addToSubCommands(command);
        command.execute();
    }

    protected void executeRemoveReference(Entity entity, String propertyName, Entity entityToRemove) {
        var command = new RemoveEntityCommandImpl();
        command.setEntity(entity);
        command.setEntityToRemove(entityToRemove);
        command.setPropertyName(propertyName);

        addToSubCommands(command);
        command.execute();
    }

}

File: simulators/de.unistuttgart.iste.sqa.mpw.framework.cpp/core/src/impl/CompositeCommandBase.cpp
#include "CompositeCommandBase.h"

#include "AddEntityCommandImpl.h"
#include "RemoveEntityCommandImpl.h"

using namespace basetypes;

namespace commands {
void CompositeCommandBase::undo() {
    auto& subCommands = getSubCommands();
    for (auto iter = subCommands.rbegin(); iter != subCommands.rend(); ++iter) {
        auto& subCommand = *iter;
        subCommand.undo();
    }
}

void CompositeCommandBase::redo() {
    for (auto& subCommand : getSubCommands()) {
        subCommand.redo();
    }
}

void CompositeCommandBase::executeSetProperty(std::shared_ptr<Entity> entity, unsigned featureKey, Any oldValue,
                                              Any newValue) {
    internalExecuteSetProperty(entity, featureKey, oldValue, newValue);
}

void CompositeCommandBase::executeSetProperty(std::shared_ptr<Entity> entity, unsigned featureKey,
                                              std::shared_ptr<Entity> oldValue, std::shared_ptr<Entity> newValue) {
    internalExecuteSetProperty(entity, featureKey, oldValue.get(), newValue.get());
}

SetPropertyCommandImpl& CompositeCommandBase::internalExecuteSetProperty(std::shared_ptr<Entity> entity, unsigned featureKey,
                                                                         Any oldValue, Any newValue) {
    auto command = std::make_shared<SetPropertyCommandImpl>();
    command->setEntity(entity);
    command->setOldValue(oldValue);
    command->setNewValue(newValue);
    command->setFeatureKey(featureKey);

    addToSubCommands(command);
    command->execute();
    return *command;
}

void CompositeCommandBase::executeAddReference(std::shared_ptr<Entity> entity, unsigned featureKey,
                                               std::shared_ptr<Entity> entityToAdd) {
    auto command = std::make_shared<AddEntityCommandImpl>();
    command->setEntity(entity);
    command->setEntityToAdd(entityToAdd);
    command->setFeatureKey(featureKey);

    addToSubCommands(command);
    command->execute();
}

void CompositeCommandBase::executeRemoveReference(std::shared_ptr<Entity> entity, unsigned featureKey,
                                                  std::shared_ptr<Entity> entityToRemove) {
    auto command = std::make_shared<RemoveEntityCommandImpl>();
    command->setEntity(entity);
    command->setEntityToRemove(entityToRemove);
    command->setFeatureKey(featureKey);

    addToSubCommands(command);
    command->execute();
}

}
